README.md
# koatty_payload
Payload parser middleware for koatty.

# 使用
-----

Koatty框架默认集成了该中间件，无需引入即可使用。

项目middleware配置 config/middleware.ts:
```
// src/config/middleware.ts
export default {
  list: [], // 加载的插件列表,执行顺序按照数组元素顺序
  config: { // 插件配置
    "PayloadMiddleware": {
        "extTypes": {
            "json": ['application/json'],
            "form": ['application/x-www-form-urlencoded'],
            "text": ['text/plain'],
            "multipart": ['multipart/form-data'],
            "xml": ['text/xml']
        },
        "limit": '20mb',
        "encoding": 'utf-8',
        "multiples": true,
        "keepExtensions": true
    }
  },
};

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.4.6](https://github.com/thinkkoa/koatty_payload/compare/v1.4.5...v1.4.6) (2023-08-30)

### [1.4.5](https://github.com/thinkkoa/koatty_payload/compare/v1.4.4...v1.4.5) (2023-07-22)

### [1.4.4](https://github.com/thinkkoa/koatty_payload/compare/v1.4.2...v1.4.4) (2023-05-23)

### [1.4.2](https://github.com/thinkkoa/koatty_payload/compare/v1.4.0...v1.4.2) (2023-02-26)


### Bug Fixes

* getMetaData return type ([4b21752](https://github.com/thinkkoa/koatty_payload/commit/4b21752df937513bc7664d5dc21788866a59cbcf))

## [1.4.0](https://github.com/thinkkoa/koatty_payload/compare/v1.3.18...v1.4.0) (2023-01-13)


### Bug Fixes

* package.json ([216697d](https://github.com/thinkkoa/koatty_payload/commit/216697da644f6a9cbad79b12b59c919ed6f70414))
* upgrade deps ([62a79c5](https://github.com/thinkkoa/koatty_payload/commit/62a79c55c8b2d3210cc63c2ac3f1d539ebc7371b))

### [1.3.18](https://github.com/thinkkoa/koatty_payload/compare/v1.3.17...v1.3.18) (2022-05-26)

### [1.3.17](https://github.com/thinkkoa/koatty_payload/compare/v1.3.16...v1.3.17) (2022-03-22)


### Bug Fixes

* 拦截参数解析异常 ([5835c50](https://github.com/thinkkoa/koatty_payload/commit/5835c5081f5ce0ec5f57d907a57c3db0643d343e))

### [1.3.16](https://github.com/thinkkoa/koatty_payload/compare/v1.3.15...v1.3.16) (2022-03-14)

### [1.3.15](https://github.com/thinkkoa/koatty_payload/compare/v1.3.14...v1.3.15) (2022-03-09)

### [1.3.14](https://github.com/thinkkoa/koatty_payload/compare/v1.3.12...v1.3.14) (2022-02-21)

### [1.3.12](https://github.com/thinkkoa/koatty_payload/compare/v1.3.11...v1.3.12) (2022-02-18)

### [1.3.11](https://github.com/thinkkoa/koatty_payload/compare/v1.3.10...v1.3.11) (2022-02-10)

### [1.3.10](https://github.com/thinkkoa/koatty_payload/compare/v1.3.8...v1.3.10) (2021-11-26)

### [1.3.8](https://github.com/thinkkoa/koatty_payload/compare/v1.3.6...v1.3.8) (2021-11-23)

### [1.3.6](https://github.com/thinkkoa/koatty_payload/compare/v1.3.4...v1.3.6) (2021-11-19)


src/parse.ts
/*
 * @Author: richen
 * @Date: 2020-11-27 17:34:42
 * @LastEditors: Please set LastEditors
 * @LastEditTime: 2022-03-22 10:06:07
 * @License: BSD (3-Clause)
 * @Copyright (c) - <richenlin(at)gmail.com>
 */
import fs from "fs";
import { parse } from "querystring";
import util from "util";
import getRawBody from "raw-body";
import inflate from "inflation";
import { parseStringPromise } from "xml2js";
import { IncomingForm, BufferEncoding } from "formidable";
import { DefaultLogger as Logger } from "koatty_logger";
import { PayloadOptions } from "./index";
import onFinished from "on-finished";
import { KoattyContext } from "koatty_core";
const fsUnlink = util.promisify(fs.unlink);
const fsAccess = util.promisify(fs.access);
/**
 *
 *
 * @export
 * @param {*} ctx
 * @param {*} options
 * @returns {*}  
 */
export function Parse(ctx: KoattyContext, options: PayloadOptions) {
    const methods = ['POST', 'PUT', 'DELETE', 'PATCH', 'LINK', 'UNLINK'];
    if (methods.every((method: string) => ctx.method !== method)) {
        return Promise.resolve({});
    }
    // defaults
    const len = ctx.req.headers['content-length'];
    const encoding = ctx.req.headers['content-encoding'] || 'identity';
    if (len && encoding === 'identity') {
        options.length = ~~len;
    }
    options.encoding = options.encoding || 'utf8';
    options.limit = options.limit || '1mb';

    if (ctx.request.is(options.extTypes.form)) {
        return parseForm(ctx, options);
    }
    if (ctx.request.is(options.extTypes.multipart)) {
        return parseMultipart(ctx, options);
    }
    if (ctx.request.is(options.extTypes.json)) {
        return parseJson(ctx, options);
    }
    if (ctx.request.is(options.extTypes.text)) {
        return parseText(ctx, options);
    }
    if (ctx.request.is(options.extTypes.xml)) {
        return parseXml(ctx, options);
    }

    return Promise.resolve({});
}

/**
 * parse form
 *
 * @param {KoattyContext} ctx
 * @param {PayloadOptions} opts
 * @returns {*}  
 */
async function parseForm(ctx: KoattyContext, opts: PayloadOptions) {
    try {
        const str = await parseText(ctx, opts);
        const data = parse(str);
        return { post: data };
    } catch (error) {
        Logger.Error(error);
        return { post: {} };
    }
}

/**
 * parse multipart
 *
 * @param {KoattyContext} ctx
 * @param {PayloadOptions} opts
 * @returns {*}  
 */
function parseMultipart(ctx: KoattyContext, opts: PayloadOptions) {

    const form = new IncomingForm({
        encoding: <BufferEncoding>opts.encoding,
        multiples: opts.multiples,
        keepExtensions: opts.keepExtensions,
    });

    let uploadFiles: any = null;
    onFinished(ctx.res, () => {
        if (!uploadFiles) {
            return;
        }
        Object.keys(uploadFiles).forEach((key: string) => {
            fsAccess(uploadFiles[key].path).then(() => fsUnlink(uploadFiles[key].path)).catch((err) => Logger.Error(err));
        });
    });
    return new Promise((resolve, reject) => {
        form.parse(ctx.req, function (err, fields, files) {
            if (err) {
                // return reject(err);
                Logger.Error(err);
                return resolve({ post: {}, file: {} });
            }
            uploadFiles = files;
            return resolve({
                post: fields,
                file: files
            });
        });
    });
}

/**
 * parse json
 *
 * @param {KoattyContext} ctx
 * @param {PayloadOptions} opts
 * @returns {*}  
 */
async function parseJson(ctx: KoattyContext, opts: PayloadOptions) {
    try {
        const str = await parseText(ctx, opts);
        const data = JSON.parse(str);
        return { post: data };
    } catch (error) {
        Logger.Error(error);
        return { post: {} };
    }
}

/**
 * parse text
 *
 * @param {KoattyContext} ctx
 * @param {PayloadOptions} opts
 * @returns {*}  {Promise<string>}
 */
function parseText(ctx: KoattyContext, opts: PayloadOptions): Promise<string> {
    return new Promise((resolve, reject) => {
        getRawBody(inflate(ctx.req), opts, function (err: any, body: string) {
            if (err) {
                // reject(err);
                Logger.Error(err);
                return resolve("");
            }
            resolve(body);
        });
    });
}

/**
 * parse xml
 *
 * @param {KoattyContext} ctx
 * @param {PayloadOptions} opts
 * @returns {*}  
 */
async function parseXml(ctx: KoattyContext, opts: PayloadOptions) {
    try {
        const str = await parseText(ctx, opts);
        const data = await parseStringPromise(str);
        return { post: data };
    } catch (error) {
        Logger.Error(error);
        return { post: {} };
    }
}

src/index.ts
/*
 * @Author: richen
 * @Date: 2020-11-27 17:07:25
 * @LastEditors: Please set LastEditors
 * @LastEditTime: 2023-02-26 13:51:51
 * @License: BSD (3-Clause)
 * @Copyright (c) - <richenlin(at)gmail.com>
 */
import Koa from "koa";
import * as helper from "koatty_lib";
import { Koatty, KoattyContext, KoattyNext } from "koatty_core";
import { DefaultLogger as logger } from "koatty_logger";
import { Parse } from "./parse";

/**
 *
 *
 * @interface DefaultOptions
 */
export interface PayloadOptions {
    extTypes: {
        json: string[],
        form: string[],
        text: string[],
        multipart: string[],
        xml: string[],
    };
    limit: string;
    encoding: BufferEncoding;
    multiples: boolean;
    keepExtensions: boolean;
    length?: number;
}

/** @type {*} */
const defaultOptions: PayloadOptions = {
    extTypes: {
        json: ['application/json'],
        form: ['application/x-www-form-urlencoded'],
        text: ['text/plain'],
        multipart: ['multipart/form-data'],
        xml: ['text/xml']
    },
    limit: '20mb',
    encoding: 'utf-8',
    multiples: true,
    keepExtensions: true,
};

/**
 *
 *
 * @export
 * @param {PayloadOptions} options
 * @param {*} app Koatty or Koa instance
 */
export function Payload(options: PayloadOptions, app: Koatty): Koa.Middleware {
    options = { ...defaultOptions, ...options };
    return async (ctx: KoattyContext, next: KoattyNext) => {
        /**
         * request body parser
         *
         * @param {any} name
         * @param {any} value
         * @returns
         */
        helper.define(ctx, 'bodyParser', async function (): Promise<any> {
            let body = ctx.getMetaData("_body")[0];
            if (!helper.isEmpty(body)) {
                return body;
            }
            try {
                const res = await Parse(ctx, options);
                body = res || {};
                ctx.setMetaData("_body", body);
                return body;
            } catch (err) {
                logger.Error(err);
                return {};
            }
        });

        /**
         * queryString parser
         *
         * @param {any} name
         * @param {any} value
         * @returns
         */
        helper.define(ctx, 'queryParser', function (): any {
            let query = ctx.getMetaData("_query")[0];
            if (!helper.isEmpty(query)) {
                return query;
            }
            query = { ...(ctx.query), ...(ctx.params || {}) };
            ctx.setMetaData("_query", query);
            return query;
        });

        return next();
    };
}